pkg/kstatus/observe/reader/cached_reader.go

```diff
@@ -0,0 +1,207 @@
+// finding all combinations of GroupVersionKind and namespace referenced by the provided
+// identifiers. This list is then expanded to include any known generated resource types.
+type CachedObserverReader struct {
+	sync.RWMutex
+
+	reader client.Reader
+
+	mapper meta.RESTMapper
+
+	gns []gvkNamespace
+
+	cache map[gvkNamespace]unstructured.UnstructuredList
+}
+
+// gvkNamespace contains information about a GroupVersionKind and a namespace.
+type gvkNamespace struct {
+	GVK       schema.GroupVersionKind
+	Namespace string
+}
+
+// Get looks up the resource identified by the key and the object GVK in the cache. If the needed combination
+// of GVK and namespace is not part of the cache, that is considered an error.
+func (c *CachedObserverReader) Get(_ context.Context, key client.ObjectKey, obj *unstructured.Unstructured) error {
+	c.RLock()
+	defer c.RUnlock()
+	gvk := obj.GetObjectKind().GroupVersionKind()
+	mapping, err := c.mapper.RESTMapping(gvk.GroupKind())
+	if err != nil {
+		return err
+	}
+	gn := gvkNamespace{
+		GVK:       gvk,
+		Namespace: key.Namespace,
+	}
+	cacheList, found := c.cache[gn]
+	if !found {
+		return fmt.Errorf("GVK %s and Namespace %s not found in cache", gvk.String(), gn.Namespace)
+	}
+	for _, u := range cacheList.Items {
+		if u.GetName() == key.Name {
+			obj.Object = u.Object
+			return nil
+		}
+	}
+	return errors.NewNotFound(mapping.Resource.GroupResource(), key.Name)
+}
+
+// ListNamespaceScoped lists all resource identifier by the GVK of the list, the namespace and the selector
+// from the cache. If the needed combination of GVK and namespace is not part of the cache, that is considered an error.
+func (c *CachedObserverReader) ListNamespaceScoped(_ context.Context, list *unstructured.UnstructuredList, namespace string, selector labels.Selector) error {
+	c.RLock()
+	defer c.RUnlock()
+	gvk := list.GroupVersionKind()
+	gn := gvkNamespace{
+		GVK:       gvk,
+		Namespace: namespace,
+	}
+
+	cacheList, found := c.cache[gn]
+	if !found {
+		return fmt.Errorf("GVK %s and Namespace %s not found in cache", gvk.String(), gn.Namespace)
+	}
+
+	var items []unstructured.Unstructured
+	for _, u := range cacheList.Items {
+		if selector.Matches(labels.Set(u.GetLabels())) {
+			items = append(items, u)
+		}
+	}
+	list.Items = items
+	return nil
+}
+
+// ListClusterScoped lists all resource identifier by the GVK of the list and selector
+// from the cache. If the needed combination of GVK and namespace (which for clusterscoped resources
+// will always be the empty string) is not part of the cache, that is considered an error.
+func (c *CachedObserverReader) ListClusterScoped(ctx context.Context, list *unstructured.UnstructuredList, selector labels.Selector) error {
+	return c.ListNamespaceScoped(ctx, list, "", selector)
+}
+
+// Sync loops over the list of gvkNamespace we know of, and uses list calls to fetch the resources.
+// This information populates the cache.
```

Comment:
Why have this in the interface at all - i.e., why not call a private sync at the top of the other methods?   

Or why not have a caching reader maintain it's own polling/caching loop rather than have some observer do the maintenance?

Reply:
So since the whole framework is based around polling with a certain frequency, having the framework call `Sync` before every polling cycle means we get state from the cluster that is as up-to-date as possible, and we only need to do one LIST call per GVK and namespace combination per cycle. 

If we do sync as part of each observer, we either end up with more sync calls than necessary, or the observers will need to be aware of the polling cycle. I was trying to avoid this, partly because it allows the observers to be reused in an implementation based on a watch instead of polling.

If the reader has its own polling loop, it will not be aligned with the framework cycle. It seems to me that would make things more complicated.